sycl/source/detail/sycl_mem_obj_t.hpp

```diff
@@ -176,7 +176,7 @@ class __SYCL_EXPORT SYCLMemObjT : public SYCLMemObjI {
   bool canReuseHostPtr(void *HostPtr, const size_t RequiredAlign) {
     bool Aligned =
         (reinterpret_cast<std::uintptr_t>(HostPtr) % RequiredAlign) == 0;
-    return !MHostPtrReadOnly && (Aligned || useHostPtr());
+    return Aligned || useHostPtr();
```

Comment:
As discussed offline, I think the name of this function is vague. What does it mean to "reuse" a pointer? With these semantics it would mean that we are allowed to read from it, but not whether or not we can write to it. I worry that someone else might end up using this function as a "reuse means any use of the pointer" and end up making the same mistakes again as we tried to solve with `!MHostPtrReadOnly &&`. On the other hand, after the shadow copy has been created, HostPtr _can_ be reused.

Personally, I would have preferred we went with the stronger "reuse" guarantee here and instead made `else if (MHostPtrReadOnly)` in the two ctors that need it, but it is a preference based on the above worries.